### ISSUE TITLE
gather-errno-name.sh: prefix stderr with name of source?

### ISSUE THREAD

**Issue by mentalisttraceur** [mentalisttraceur/errnoname#8]
Every time I've had a failure in `gather-errno-names.sh`, my first question was "which of these has failed?"

So the short version of the question: is it worth it to add a function to prefix the standard error of every one of the source commands?

**Comment by mentalisttraceur** [comment]
Unfortunately, the shell logic for piping standard error is... convoluted and byzantine. Off the top of my head, it looks something like this:

```sh
{ some-command 2>&1 1>&3 | sed 's/^/prefix: /' 1>&2; } 3>&1
```

**Comment by mentalisttraceur** [comment]
That's actually not too bad, assuming I actually got it right.

Disclaimer: the above also lost the exit code of `some-command`! It doesn't matter for this usecase because in `gather-errno-names.sh` we already swallow most exit statuses because of extensive piping anyway, but in the general case that's worth noticing as a regression of the above relative to just invoking `some-command`.

Here's what it looks like to fix that (portably, not relying on things like `PIPESTATUS` or whatever; although this might still not be maximally portable for some `/bin/sh`, I'd have to double check things like the reliablity of double quotes both outside and inside backticks):

```sh
(exit "`{ some-command 2>&1 1>&3; echo "$?" 1>&4 | sed 's/^/prefix: /' 1>&2; } 4>&1`") 3>&1
```

**Comment by mentalisttraceur** [comment]
Anyway, the point is, that's not very readable. Putting it in a function is probably not too bad:

```sh
execute_with_prefixed_errors()
{
    command=$1
    { "$command" 2>&1 1>&3 | sed "s/^/$command: /" 1>&2; } 3>&1
}
```

Because we really want to get the prefix from the command so that we can write `execute_with_prefixed_errors solaris` rather than things like `execute_with_prefixed_errors solaris solaris`.

But! If we want to retain existing functionality, then we have to deal with the fact that currently `gather-errno-names.sh solaris` calls the `solaris` function inside the script, and so we either:

1. escape the first argument before making it the prefix so that it doesn't make `sed` behave poorly,
2. ~~somehow validate the first argument before passing it to `sed` (in POSIX shell, `command -v "$command" || return 1` is probably good enough, because as far as I can tell, anything that's a valid Bourne/POSIX shell command name is also safe in the second half of a `sed` substitution)~~, or
3. have some other boilerplate to translate from the command-line argument `solaris` to `execute_with_prefixed_errors solaris`.

I think the second option is the only one where the downsides are worth the upsides, so it would look something like this:

```sh
execute_with_prefixed_errors()
{
    command=$1
    # If $command is the name of a valid shell command, it is
    # also safe in the second half of the `sed` substitution:
    command -v "$command" || return 1
    { "$command" 2>&1 1>&3 | sed "s/^/$command: /" 1>&2; } 3>&1
}
```

Edit: but actually, after some testing `command -v` is **not** sufficient for filtering out all things that would make `sed` to potentially harmful things, so it would necessarily invole more non-trivial code.

**Comment by mentalisttraceur** [comment]
Actually:

```sh
echo 'echo \;' >'/tmp/;'
chmod +x '/tmp/;'
command -v '/tmp/;'
```

That succeeds on a BusyBox ash shell I have on hand. So that's not good enough.

Broadly speaking, if someone out there has a pipeline which calls `gather-errno-names.sh "$some_variable"`, it is our job to error out on bad inputs without doing things like overwriting files or whatever.

More broadly, the responsibilities for an invocation like `some-program {{ argument }}` break down as follows:

1. The caller is responsible for correctly passing the argument into `some-program`.
2. `some-program` is responsible for not doing things that are not part of its official behavior.

So if you ultimately cause `gather-errno-name.sh` to be invoked with one of the arguments like `/w foo.txt; ...`, it is the job for `gather-errno-name.sh` to not write file `foo.txt` in the local directory. And remember that it's very possible that `gather-errno-name.sh` is getting invoked by some off-the-cuff one-liner written in an interactive shell session, or from scripts/programs/etc which might have bugs that don't validate/reject bad inputs, and which might programmatically assemble the arguments to `gather-errno-name.sh`, and which might have bugs which cause garbage characters to end up in the arguments, and so on. So we can't assume that the only way the argument would be bad would be the user who suffers the consequences willfully doing something obviously wrong.

**Comment by mentalisttraceur** [comment]
So this is probably the smallest, cleanest version of what it would ultimately have to look like:

```sh
execute_with_prefixed_errors()
{
    command=$1
    # Simple validation which will reject anything that is not a
    # valid `sh` function name, which also eliminates anything
    # that is unsafe in the second half of a `sed` substitution.
    case $command in [!a-zA-Z_]* | ?*[!a-zA-Z0-9_]*)
        printf '%s\n' "bad subcommand: $command" 1>&2
        return 1
    esac
    { "$command" 2>&1 1>&3 | sed "s/^/$command: /" 1>&2; } 3>&1
}
```

**Comment by mentalisttraceur** [comment]
I think I like this. I think I'm gonna do it. I think I've talked myself into it.

**Comment by mentalisttraceur** [comment]
Implemented in 86d6ab9fa4a297937851ed9b53931d8dc64ba8d2.
